project/src/main/ui/level-select/hookable-grade-label.gd

```diff
@@ -0,0 +1,120 @@
+Assigns this label's button, and updates the label's appearance.
+"""
+func set_button(new_button: LevelSelectButton) -> void:
+	button = new_button
+	
+	button.get_node("GradeHook").remote_path = button.get_node("GradeHook").get_path_to(self)
+	if not button.is_connected("lowlight_changed", self, "_on_LevelSelectButton_lowlight_changed"):
+		button.connect("lowlight_changed", self, "_on_LevelSelectButton_lowlight_changed")
+	if not button.is_connected("tree_exited", self, "_on_LevelSelectButton_tree_exited"):
+		button.connect("tree_exited", self, "_on_LevelSelectButton_tree_exited")
+	
+	_refresh_appearance()
+
+
+"""
+Updates the text/icon based on the level's status.
+
+If the level has been cleared, we display text corresponding to the player's grade. If the level hasn't been cleared,
+we show a lock/key/crown icon corresponding to the level's lock status.
+"""
+func _refresh_appearance() -> void:
+	if button.is_class("WorldSelectButton"):
+		var world_button: WorldSelectButton = button
+		var worst_rank := 0.0
+		for rank in world_button.ranks:
+			worst_rank = max(worst_rank, rank)
+		_refresh_label_text(RankCalculator.grade(worst_rank))
+	else:
+		var result := PlayerData.level_history.best_result(button.level_id)
+		if not result:
+			_refresh_status_icon()
+		elif result.compare == "-seconds":
+			_refresh_label_text(RankCalculator.grade(result.seconds_rank))
+		else:
+			_refresh_label_text(RankCalculator.grade(result.score_rank))
+	
+	visible = false if button.lowlight else true
+
+
+"""
+Updates the icon based on the level's status.
+
+We show a lock/key/crown icon corresponding to the level's lock status.
+"""
+func _refresh_status_icon() -> void:
+	match button.lock_status:
+		LevelLock.STATUS_NONE:
+			$StatusIcon.texture = null
+			$StatusIcon.modulate = Color.white
+			$StatusIcon.material.set("shader_param/black", Color("b39a8f"))
+		LevelLock.STATUS_CROWN:
+			$StatusIcon.texture = _crown_texture
+			$StatusIcon.modulate = Color("36d936")
+			$StatusIcon.material.set("shader_param/black", Color("b39a8f"))
+		LevelLock.STATUS_KEY:
+			$StatusIcon.texture = _key_texture
+			$StatusIcon.modulate = Color("36d936")
+			$StatusIcon.material.set("shader_param/black", Color("b39a8f"))
+		LevelLock.STATUS_SOFT_LOCK:
+			$StatusIcon.texture = _locked_texture
+			$StatusIcon.modulate = Color("666666")
+			$StatusIcon.material.set("shader_param/black", Color("808080"))
+		LevelLock.STATUS_HARD_LOCK:
+			$StatusIcon.texture = _locked_texture
+			$StatusIcon.modulate = Color("444444")
+			$StatusIcon.material.set("shader_param/black", Color("808080"))
+	
+	$GradeLabel.visible = false
+	$StatusIcon.visible = true
+
+
+"""
+Updates the text based on the player's grade on a level.
+"""
+func _refresh_label_text(new_label_text: String) -> void:
+	$GradeLabel.text = new_label_text
+	var font: DynamicFont = $GradeLabel.get("custom_fonts/font")
+	var font_color: Color = $GradeLabel.get("custom_colors/font_color")
+	var outline_harshness := 0.2
```

Comment:
Can inline this variable; it doesn't varry

Reply:
Fixed "M", "SSS" ranks to have brighter outlines. This helps them stand out more. Renamed to "outline_darkness" which is less ambiguous.